### PR TITLE
Add FXIOS-4568 [v105] Wallpaper Onboarding UI skeleton

### PR DESCRIPTION
**Pull request by thatswinnie**
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-4568)
#11307

### CODE REVIEW

File: Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewController.swift [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return viewModel.wallpaperCellModels.count
+    }
+
+    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: WallpaperCollectionViewCell.cellIdentifier,
+                                                            for: indexPath) as? WallpaperCollectionViewCell,
+                let cellViewModel = viewModel.wallpaperCellModels[safe: indexPath.row] else {
+            return UICollectionViewCell()
+        }
```
Comment (electricRGB) [inline Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewController.swift:126]:
```suggestion
        guard let cell = collectionView.dequeueReusableCell(withReuseIdentifier: WallpaperCollectionViewCell.cellIdentifier,
                                                            for: indexPath) as? WallpaperCollectionViewCell,
                let cellViewModel = viewModel.wallpaperCellModels[safe: indexPath.row] 
        else { return UICollectionViewCell() }
```
nit: guard/else on separate line. Also, identation. I think.

File: Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewModel.swift [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import Foundation
+
+class WallpaperSelectorViewModel {
+
+    enum WallpaperSelectorLayout: Equatable {
+        case compact
+        case regular
+
+        // The maximum number of items to display in the whole section
+        var maxItemsToDisplay: Int {
```
Comment (electricRGB) [inline Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewModel.swift:14]: nits for this file: return on the same line as the case as it's small and one line. Maybe. Maybe not. Up to you.

File: Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            case .regular:
+                return 5
+            }
+        }
+    }
+
+    private var wallpaperManager: WallpaperManagerInterface
+    var openSettingsAction: (() -> Void)
+    var sectionLayout: WallpaperSelectorLayout = .compact // We use the compact layout as default
+    var wallpaperCellModels: [WallpaperCellViewModel] = []
```
Comment (electricRGB) [inline Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewModel.swift:47]:
```suggestion
    var wallpaperCellModels = [WallpaperCellViewModel]()
```
nit: we haven't made a style decision. But, maybe?

File: Client/Frontend/Home/HomepageViewController.swift [hunk 4]
```diff
@@ -289,6 +298,27 @@ class HomepageViewController: UIViewController, HomePanel {
...
 
+    private func displayWallpaperSelector() {
+        guard let wallpaperVersion: WallpaperVersion = featureFlags.getCustomState(for: .wallpaperVersion),
+              wallpaperVersion == .v2,
+              featureFlags.isFeatureEnabled(.wallpaperOnboardingSheet, checking: .buildOnly)
+        else { return }
+
+        self.dismissKeyboard()
+
+        let viewModel = WallpaperSelectorViewModel(wallpaperManager: WallpaperManager(), openSettingsAction: {
```
[the remark this answers is not in the document]
Comment (thatswinnie) [reply Client/Frontend/Home/HomepageViewController.swift:309]: If it was a completion argument I'd agree but for `openSettingsAction` I fear that people might assume it's a completion block if I change it trailing closure syntax.

File: Client/BottomSheet/BottomSheetViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    public var backgroundColor: UIColor
+    public var sheetLightThemeBackgroundColor: UIColor
+    public var sheetDarkThemeBackgroundColor: UIColor
+
+    public init() {
+        cornerRadius = 8
+        animationTransitionDuration = 0.3
+        backgroundColor = .clear
+        sheetLightThemeBackgroundColor = UIColor.Photon.LightGrey10
+        sheetDarkThemeBackgroundColor = UIColor.Photon.DarkGrey40
```
[the remark this answers is not in the document]
Comment (yoanarios) [reply Client/BottomSheet/BottomSheetViewModel.swift]: I had a similar comment but once again I'm to slow :)

File: Client/Frontend/Home/Wallpapers/v2/UI/WallpaperCollectionViewCell.swift [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+    // MARK: - UI Element
+    private lazy var imageView: UIImageView = .build { imageView in
+        imageView.contentMode = .scaleAspectFill
+        imageView.clipsToBounds = true
+    }
+
+    private lazy var borderView: UIView = .build { borderView in
+        borderView.layer.cornerRadius = 10
+        borderView.layer.borderWidth = 1
```
Comment (yoanarios) [inline Client/Frontend/Home/Wallpapers/v2/UI/WallpaperCollectionViewCell.swift:17]: Can we create struct UX and add the values

File: Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewController.swift [hunk 7]
```diff
@@ -0,0 +1,249 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import UIKit
+
+class WallpaperSelectorViewController: UIViewController {
+
+    private struct UX {
+        static let cardWidth: CGFloat = UIDevice().isTinyFormFactor ? 88 : 97
```
Comment (yoanarios) [inline Client/Frontend/Home/Wallpapers/v2/UI/WallpaperSelectorViewController.swift:10]: Nice, I will use this extension for the onboarding views

File: Client/Frontend/Intro/IntroViewModel.swift [hunk 8]
```diff
@@ -45,7 +45,11 @@ struct IntroViewModel: OnboardingViewModelProtocol {
     }
 
     var enabledCards: [InformationCards] {
-        return [.welcome, .wallpapers, .signSync]
+        guard let wallpaperVersion: WallpaperVersion = featureFlags.getCustomState(for: .wallpaperVersion) else {
```
Comment (yoanarios) [inline Client/Frontend/Intro/IntroViewModel.swift:48]: I will have to update to use a another feature flag to hide the card but I will keep the logic

File: Client/Frontend/Strings.swift [hunk 9]
```diff
@@ -725,6 +725,31 @@ extension String {
             value: "Sign Up and Log In",
             comment: "Describes an action on the sync onboarding page in our Onboarding screens. This string will be on a button so user can sign up or login directly in the onboarding.",
             lastUpdated: .v102)
+        public static let WallpaperSelectorTitle = MZLocalizedString(
+            "Onboarding.Wallpaper.Title.v105",
+            value: "Try a splash of color",
+            comment: "Title for the wallpaper onboarding modal displayed on top of the homepage. This describes to the user that they can choose different wallpapers.",
+            lastUpdated: .v105)
```
Comment (yoanarios) [inline Client/Frontend/Strings.swift:732]: Should we tag for v106? because the string freeze for v105 was already done

Comment (thatswinnie) [reply Client/Frontend/Strings.swift:732]: Good point! Will do.

**Review by yoanarios — APPROVED** [review]

**Comment by thatswinnie** [comment]
Screen Shot 2022-08-19 at 10 03 48

**Pull request merged**